python/opmonlib/publisher_base.py

```diff
@@ -0,0 +1,178 @@
+import sys
+from abc import ABC, abstractmethod
+
+from google.protobuf.descriptor import FieldDescriptor
+from google.protobuf.message import Message as Msg
+from google.protobuf.timestamp_pb2 import Timestamp
+
+from opmonlib.conf import OpMonConf
+from opmonlib.opmon_entry_pb2 import OpMonEntry, OpMonId, OpMonValue
+from opmonlib.utils import (
+    LogLevelError,
+    logging_log_levels,
+    oks_log_levels,
+    oks_to_logging_map,
+)
+
+
+class OpMonPublisherBase(ABC):
+    """Base class for OpMon publishers."""
+
+    @abstractmethod
+    def __init__(self) -> None:
+        """Construct the publisher."""
+        self.ts = Timestamp()
+        self.substructure = None
+        pass
+
+    def __post_init__(self) -> None:
+        """Perform post-init checks."""
+        if not self.log:
+            err_msg = "OpMon publisher must have a logger."
+            raise (err_msg)
+        if not self.conf:
+            err_msg = "OpMon publisher must have a configuration."
+            raise AttributeError(err_msg)
+        if not isinstance(self.conf, OpMonConf):
+            err_msg = "OpMon configuration must be of type OpMonConf."
+            raise TypeError(err_msg)
+        if not self.publisher:
+            err_msg = "OpMon publisher must have a publisher"
+            raise AttributeError(err_msg)
+        if not self.default_topic:
+            err_msg = "OpMon publisher must have a default_topic"
+            raise AttributeError(err_msg)
+        return
+
+    @abstractmethod
+    def publish(
+        self,
+        message: Msg,
+        custom_origin: dict[str, str] | None = None,
+        substructure: list[str] | None = None,
```

Comment:
No, we said also the substructure has to become a member and not being an argument of the publish method.